### ISSUE TITLE
dxtbx: issue with pycbf

### ISSUE THREAD

**Issue by picca** [dials/dials#2196]
Hello, I am sill trying to obtain a working dials stack on Debian :).

first I do not know which dxtbx repository (the one from cctbx of your) is the official repository of dxtbx. I can not fill issue about dxtbx on your repository, but the cctbx is a fork of your. So I am wondering which one should be use in order to work. today it seems that both are in sync.

Now I am coming back with the string / bytes issue due to pycbf.

I packaged the latest 0.9.7 version of cbflib and use the pycbf module (which is the Python3 version), and indeed I got this kind of error

```python

    def test_beam():
        image = str(Path(__file__).parent / "phi_scan_001.cbf")
>       assert BeamFactory.imgCIF(image)

tests/test_beam.py:10: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dxtbx/model/beam.py:259: in imgCIF
    cbf_handle.read_widefile(cif_file.encode(), pycbf.MSG_DIGEST)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pycbf.cbf_handle_struct; proxy of <Swig Object of type 'handle *' at 0x7f08b6b764b0> >
filename = b'/home/experiences/instrumentation/picca/debian/science-team/dxtbx/.pybuild/cpython3_3.9_dxtbx/build/tests/phi_scan_001.cbf', headers = 8

    def read_widefile(self, filename, headers):

[...]

>       return _pycbf.cbf_handle_struct_read_widefile(self, filename, headers)
E       TypeError: in method 'cbf_handle_struct_read_widefile', argument 2 of type 'char *'

/usr/lib/python3/dist-packages/pycbf.py:2086: TypeError
```

the encoding is not required anymore. it seems that pycbf expect now real python3 string and not byte.

I am a little bit confuse, and I would like to know how you make all this work together.

thanks for considering.

Frederic

**Comment by ndevenish** [comment]
> today it seems that both are in sync.

Yes, they are the same. For historical reasons, all the issues and PRs go against https://github.com/cctbx/dxtbx

> I packaged the latest 0.9.7 version of cbflib

Ah, I don't think any of us were aware that a 0.9.7 had been released. We were told that CBFlib development was moving to https://github.com/cbflib/cbflib which doesn't have a release, or any new commits left. DIALS has certainly never been built against 0.9.7 from here, it still uses 0.9.6. In fact, the 0.9.7 development branch PR appears to still be open and have more recent commits than the release date https://github.com/cbflib/cbflib/pull/1 ?

DIALS as-is used to have CBFlib to be built with `SWIG_PYTHON_STRICT_BYTE_CHAR`. In fact, to get around this and to release the conda-forge build of DIALS, we split off the pycbf parts into https://github.com/dials/pycbf, and added the python3 compatibility changes - this meant that pycbf could accept transparently bytes and strings. We were told that it was of prime importance that every existing use case would build and work as-is with CBFlib 0.9.7, so I am surprised that it doesn't work automatically.

In any case, probably there is work to be done to reconcile 0.9.7 CBFlib into DIALS.

**Comment by picca** [comment]
> &gt; I packaged the latest 0.9.7 version of cbflib

I am using this repository https://github.com/yayahjb/cbflib which contain a 0.9.7 tag.

It would be great effectively that I know wich repository is the right one. Thsi way I should avoid this kind of issue.

> Ah, I don&#039;t think any of us were aware that a 0.9.7 had been released. We were told that CBFlib development was moving to https://github.com/cbflib/cbflib which doesn&#039;t have a release, or any new commits left. DIALS has certainly never been built against 0.9.7 from here, it still uses 0.9.6. In fact, the 0.9.7 development branch PR appears to still be open and have more recent commits than the release date https://github.com/cbflib/cbflib/pull/1 ?

> DIALS as-is used to have CBFlib to be built with `SWIG_PYTHON_STRICT_BYTE_CHAR`. In fact, to get around this and to release the conda-forge build of DIALS, we split off the pycbf parts into https://github.com/dials/pycbf, and added the python3 compatibility changes - this meant that pycbf could accept transparently bytes and strings. We were told that it was of prime importance that every existing use case would build and work as-is with CBFlib 0.9.7, so I am surprised that it doesn&#039;t work automatically.

MAybe I have the wronf cbflib :(

> In any case, probably there is work to be done to reconcile 0.9.7 CBFlib into DIALS.

It would be great to have only one cbflib to care about.

Cheers

**Comment by ndevenish** [comment]
No, you have the right CBFlib. It's just DIALS is currently only built against the older 0.9.6 version, with cctbx' cbflib_adaptbx.

We did make changes in DIALS so that it could be added as a conda-forge package, but to work without cbflib_adaptbx (which wasn't present in the cctbx package), we added changes to let pycbf work without them. That, and tidying up the build so that pycbf could be stand-alone installed, was why https://github.com/dials/pycbf came about.

I can't find a way to build cctbx with the latest cbflib, all of the cctbx bootstrap methods I knew appear to be broken (either dead URLS or tqdm errors from miniforge), so I'm building a fresh DIALS to see if the new CBFlib works with it now.

In fact, looking at the code, it looks like instead of resolving the byte/str issue the same way we did, it looks like he is manually injecting `SWIG_PYTHON_STRICT_BYTE_CHAR` into the generated build to work around the issue?
https://github.com/yayahjb/cbflib/blob/5232590ec6cc05bb2705acc3413ea598263854da/pycbf/make_pycbf.py#L633

**Comment by ndevenish** [comment]
In fact, for a fresh dials bootstrap, manually updating to CBFlib 0.9.7 works perfectly:
```
git clone git@github.com:dials/dials modules/dials
git clone https://github.com/yayahjb/cbflib -b CBFlib-0.9.7 modules/cbflib
python3 modules/dials/installer/bootstrap.py base update build
source ./dials
pytest -n auto modules/dxtbx --regression
```
So, whatever the cctbx parts of the cbflib build are running appear to be compiling it the same way, it appears to be your build that is different.

Perhaps it is requiring `SWIG_PYTHON_STRICT_BYTE_CHAR`. On our intermediate pycbf repository I added a commit that let anything accepting str or bytes to accept both: https://github.com/dials/pycbf/pull/4/files, or there was my PR https://github.com/yayahjb/cbflib/pull/19 to resolve this directly, but it looks like the approach I noted above was used instead. Historically, there has been opposition to switching to non-strict pycbf, but that might be resolved now that it's been forced on for.... parts of the build. Still, migrating dxtbx to non-strict is probably some medium body of work.

**Comment by ndevenish** [comment]
In any case, since this seems to be a problem in your custom build, I'm converting this to a discussion rather than an issue.
